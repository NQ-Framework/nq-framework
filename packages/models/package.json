{
  "name": "@nqframework/models",
  "version": "4.2.0",
  "description": "Data Transfer Objects for the NQ Framework",
  "main": "build/index.js",
  "module": "build/index.es.js",
  "jsnext:main": "build/index.es.js",
  "scripts": {
    "build": "tsc",
    "test": "jest",
    "start:dev": "tsc --watch",
    "test:watch": "jest --watch",
    "test:coverage": "jest --coverage",
    "format": "prettier --write \"src/**/*.ts\"",
    "lint": "tslint -p tsconfig.json",
    "prepublishOnly": "npm test && npm run lint && npm run build",
    "preversion": "npm run lint",
    "version": "npm run format && git add -A src",
    "postversion": "git push && git push --tags"
  },
  "files": [
    "build",
    "LICENCE"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/nq-framework/models"
  },
  "keywords": [],
  "author": "Milos Spasojevic",
  "license": "MIT",
  "devDependencies": {
    "@types/jest": "^25.2.3",
    "jest": "^25.5.4",
    "prettier": "^2.1.2",
    "rollup": "^2.33.1",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-peer-deps-external": "^2.2.4",
    "rollup-plugin-typescript2": "^0.26.0",
    "ts-jest": "^25.5.1",
    "tslint": "^6.1.3",
    "tslint-config-prettier": "^1.18.0",
    "typescript": "^3.9.7"
  },
  "dependencies": {
    "@irreal/decorator-mapper": "^1.0.21",
    "reflect-metadata": "^0.1.13"
  }
}
