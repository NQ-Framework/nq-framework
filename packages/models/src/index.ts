export { Organization } from "./organization/organization";
export { OrganizationMember } from "./organization/organization-member";
export { DataSource } from "./data-fetching/data-source";
export { MsSqlConfiguration } from "./data-fetching/ms-sql-configuration";
export { ApiConfiguration } from "./data-fetching/api-configuration";
export { ScheduledJob } from "./scheduled-jobs/scheduled-job";
export { TriggeredJob } from "./triggered-jobs/triggered-job";
export { ConfigurationInterface } from "./jobs/configuration-interface";
export { LogConfiguration } from "./jobs/log-configuration";
export {
  StoredProcedureConfiguration,
  ProcedureParameter,
} from "./jobs/stored-procedure-configuration";
export { Action } from "./workflow/actions/action";
export { ActionInstance } from "./workflow/actions/action-instance";
export { ActionResult } from "./workflow/actions/action-result";
export { ActionHandler } from "./workflow/actions/action-handler";
export { ActionLink } from "./workflow/actions/action-link";
export { Workflow } from "./workflow/workflow";
export { WorkflowExecutionData } from "./workflow/workflow-execution-data";
export { WorkflowExecutionResult } from "./workflow/workflow-execution-result";
export { WorkflowExecutionContext } from "./workflow/workflow-execution-context";
