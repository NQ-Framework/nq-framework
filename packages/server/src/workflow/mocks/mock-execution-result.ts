import { WorkflowExecutionResult } from '@nqframework/models';

export const mockExecutionResult: WorkflowExecutionResult = {
    data: {
        mockProp: "mock value"
    },
    finalOutput: {
        mockFinalOutputProp: "mock final value"
    }
}