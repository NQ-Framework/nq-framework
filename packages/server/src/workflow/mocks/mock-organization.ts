import { Organization } from '@nqframework/models';

export const mockOrganization: Organization = {
    name: "mock org id",
    members: [{
        displayName: "mock display name",
        email: "mock@email.com",
        role: "admin",
        status: "accepted",
        uid: "mock user id"
    }],
    memberIds: ["mock user id"],
    address1: "addr1",
    address2: "addr2",
    country: "ctry",
    dataCredentials: [],
    email: "mock-org@email.com",
    serviceAccounts: [],
    dataSources: []
}