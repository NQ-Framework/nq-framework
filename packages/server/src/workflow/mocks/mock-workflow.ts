import { Workflow, WorkflowTriggerInstance } from '@nqframework/models';

export const mockWorkflow: Workflow = {
  id: 'mock id',
  isActive: true,
  name: 'mock workflow',
  organizationId: 'mock org id',
  actionInstances: [
    {
      name: 'mock action instance id',
      isEnabled: true,
      editorConfig: {
        color: '#000000',
        x: 100,
        y: 100,
      },
      configuration: {
        input: [],
      },
      action: {
        id: 'mock action id',
        isEnabled: true,
        name: 'mock action',
        path: 'mock-path',
        version: 1,
        hasDefaultPort: true,
      },
    },
    {
      name: 'mock action instance id 2',
      isEnabled: true,
      editorConfig: {
        color: '#000000',
        x: 100,
        y: 100,
      },
      configuration: {
        input: [],
      },
      action: {
        id: 'mock action id',
        isEnabled: true,
        name: 'mock action',
        path: 'mock-path',
        version: 1,
        hasDefaultPort: true,
      },
    },
  ],
  actionLinks: [
    {
      fromName: 'mock action instance id',
      toName: 'mock action instance id 2',
      isEnabled: true,
    },
  ],
  triggers: [
    {
      actions: ['mock action instance id'],
      editorConfig: {
        color: '#000000',
        x: 100,
        y: 100,
      },
      id: '123',
      input: [
        { name: 'verb', value: ['GET'] },
        { name: 'endpoint', value: 'test endpoint' },
      ],
      output: [],
      type: 'api',
    } as WorkflowTriggerInstance,
  ],
};
