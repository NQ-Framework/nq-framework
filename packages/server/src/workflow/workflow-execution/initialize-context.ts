import { WorkflowExecutionContext } from '@nqframework/models';

export function initializeContext(
  context: WorkflowExecutionContext,
): WorkflowExecutionContext {
  const workflow = context.workflow;
  const actionInstances = workflow.actionInstances
    ? workflow.actionInstances.filter((ai) => ai.isEnabled)
    : [];

  if (actionInstances.length === 0) {
    throw new Error(
      'Execution context cannot be initialized with zero actions',
    );
  }

  let actionLinks = workflow.actionLinks;
  if (!actionLinks) {
    actionLinks = [];
  } else {
    actionLinks = actionLinks.filter((al) =>
      al.isEnabled &&
      actionInstances.some((ai) => ai.name === al.fromName) &&
      actionInstances.some((ai) => ai.name === al.toName)
    );
  }
  return {
    ...context,
    workflow: { ...workflow, actionInstances, actionLinks },
  };
}
