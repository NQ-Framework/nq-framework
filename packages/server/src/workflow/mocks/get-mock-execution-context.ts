import { WorkflowExecutionContext } from '@nqframework/models';
import { mockOrganization } from './mock-organization';
import { mockWorkflow } from './mock-workflow';

export const getMockExecutionContext = (): WorkflowExecutionContext =>
  JSON.parse(
    JSON.stringify({
      actions: {
        'mock name': {
          properties: [{ name: 'mock prop', value: 'mock prop val' }],
          values: [{ name: 'mock value ', value: 'mock value val' }],
        },
      },
      input: { mockInput: 'mock input value' },
      id: 'mock execution context id',
      startedBy: 'mock user id',
      organization: mockOrganization,
      isRunning: true,
      stack: [],
      startTime: new Date('2020-11-20'),
      triggerInput: [{ name: 'trigInput', value: 'mock trigger input value' }],
      triggerOutput: { output: 'mock trigger output' },
      workflow: mockWorkflow,
    } as WorkflowExecutionContext),
  ) as WorkflowExecutionContext;
