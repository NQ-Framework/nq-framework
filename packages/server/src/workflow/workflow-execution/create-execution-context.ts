import { PropertyValue, Workflow, WorkflowExecutionContext } from "@nqframework/models";
import { evaluateProperties, reducePropertyValuesToObject } from '../../core/utils';
import { initializeContext } from './initialize-context';

export async function createExecutionContext(input: PropertyValue[], workflow: Workflow): Promise<WorkflowExecutionContext> {
    let context: WorkflowExecutionContext = {
        isRunning: true,
        startTime: new Date(),
        stack: [],
        actions: {},
        input: {},
        triggerInput: input,
        workflow: workflow,
    };
    context.input = reducePropertyValuesToObject(
        await evaluateProperties(input, context),
    );

    context = initializeContext(context);

    return context;
}

