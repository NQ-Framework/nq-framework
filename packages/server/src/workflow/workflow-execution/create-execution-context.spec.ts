import { mockWorkflow } from '../mocks/mock-workflow';
import { createExecutionContext } from './create-execution-context';
import {
  evaluateProperties,
} from '../../core/utils';
import { initializeContext } from './initialize-context';
import { reducePropertyValuesToObject } from '@nqframework/models';
import { mockOrganization } from '../mocks/mock-organization';

jest.mock('@nqframework/models', () => ({
  ...(jest.requireActual('@nqframework/models') as any),
  reducePropertyValuesToObject: jest.fn(),
}));
jest.mock('../../core/utils', () => ({
  ...(jest.requireActual('../../core/utils') as any),
  evaluateProperties: jest.fn(),
}));

jest.mock('./initialize-context');

const FIXED_SYSTEM_TIME = '2020-11-10T00:00:00.000Z';

describe('createExecutionContext', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers('modern');
    jest.setSystemTime(Date.parse(FIXED_SYSTEM_TIME));
    (initializeContext as jest.Mock).mockImplementation((context) => {
      return context;
    });
  });
  it('creates the context', async () => {
    const result = await createExecutionContext([], mockWorkflow, mockOrganization);
    expect(result).toBeDefined();
    expect(result.startTime.toISOString()).toEqual(FIXED_SYSTEM_TIME);
  });

  it('evaluate static properties', async () => {
    (reducePropertyValuesToObject as jest.Mock).mockImplementation(() => {
      return {
        'test prop': 'test value',
      };
    });
    const result = await createExecutionContext(
      [{ name: 'test prop', value: 'test value' }],
      { ...mockWorkflow },
      mockOrganization
    );
    expect(result.input['test prop']).toEqual('test value');
    expect(reducePropertyValuesToObject).toHaveBeenCalledTimes(1);
    expect(evaluateProperties).toHaveBeenCalledTimes(1);
  });
});
