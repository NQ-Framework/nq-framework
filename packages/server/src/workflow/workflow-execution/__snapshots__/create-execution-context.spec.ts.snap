// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createExecutionContext throw if workflow has no actions 1`] = `"Execution context cannot be initialized with zero actions"`;

exports[`createExecutionContext throw if workflow has no actions 2`] = `"test"`;
